build: Enable TypeScript type checking (#7680)
plotly/plotly.js: src/types/GENERATOR.md

# Type Generator Internals

The **schema-based generator** (`tasks/generate_schema_types.mjs`)
reads `plot-schema.json` and emits all the schema-derived TypeScript
types into `src/types/generated/schema.d.ts`:

- Common enum aliases (Calendar, Dash, AxisType, PatternShape, XRef, YRef,
  TransitionEasing, PlotType)
- Shared sub-interfaces (Font, ColorBar, HoverLabel, etc.)

> - 49 per-trace data interfaces (BarData, ScatterData, IndicatorData, ...)

Reviewer: ```suggestion
- Data interfaces for each trace type (BarData, ScatterData, IndicatorData, etc.)
```

    { name: 'Calendar', match: (key) => /^[xyz]?calendar$/.test(key) },
    { name: 'Dash', match: (key) => key === 'dash' },
    { name: 'AxisType', match: (key, path) => key === 'type' && /[xyz]axis\.type$/.test(path) },
    // ...
];
```

When multiple sites match an anchor (e.g. 3D scene axes vs cartesian axes
both have `xaxis.type` enumerations), the generator picks the largest
value set — the superset — so the alias is always permissive enough.
`PlotType` is special-cased: derived from `Object.keys(schema.traces)`
rather than from an attribute.

Each discovered alias is emitted as `export type Name = 'a' | 'b' | ...`
and registered in `VALUES_TO_COMMON_TYPE` so subsequent emission of any
attribute whose `values` array matches an anchor produces a reference
to the alias instead of an inlined literal union.

### Phase 1: Fingerprinting

The generator fingerprints every container subtree across all traces and
layout. Two containers are considered identical when their sorted keys
and leaf `valType`s produce the same fingerprint string.

### Phase 2: Shared interface extraction

Containers that appear at least `MIN_OCCURRENCES` (= 5) times AND have at

> least `MIN_PROPERTIES` (= 4) properties become shared interfaces (Font,

Reviewer: ```suggestion
Containers that appear at least `MIN_OCCURRENCES` times AND have at
least `MIN_PROPERTIES` properties become shared interfaces (Font,
```

Containers (no `valType`) only carry a description, so the metadata
branches no-op. Plotly's `*emphasis*` markers are preserved as-is (they
render as italics in IDE hover tooltips). Any `*/` sequences in
descriptions are escaped to prevent prematurely closing the comment.

### Output structure

```
src/types/generated/schema.d.ts
├── import { Color, ColorScale, Datum, MarkerSymbol, TypedArray } from '../lib/common'
├── Common enum types (Calendar, Dash, AxisType, PatternShape, XRef, YRef,
│                      TransitionEasing, PlotType)
├── Shared interfaces — public (Font, FontArray, ColorBar, HoverLabel, Domain,
│                                Pattern, TickFormatStops, LegendGroupTitle, ...)
├── Internal shared interfaces in `namespace _internal` (Marker, Line,
│                                                        AutoRangeOptions,
│                                                        Lighting, Stream, ErrorY)
├── Trace interfaces (ScatterData, BarData, ... — 49 traces)
├── Layout component interfaces (LayoutAxis, Legend, Scene, Annotation, etc.)
├── Layout interface
└── Animation / frames / config (AnimationOpts, Frame, Edits, ConfigBase)
```

> Regenerate with `npm run schema`.

Reviewer: ```suggestion
```

type checking. Not implemented today; consider whether the autocomplete
win is worth the cost before changing this.

## Debugging

If the schema generator emits unexpected types:

```bash
npm run schema    # regenerate and inspect schema.d.ts
npm run typecheck # see what tsc thinks
```

Inspect the schema directly:

```js
const s = require("test/plot-schema.json");
console.log(s.layout.layoutAttributes.xaxis);  // inspect layout attrs
console.log(s.traces.scatter.attributes);       // inspect trace attrs
```

## Public API re-export

`lib/index.d.ts` uses `export type * from '../src/types/generated/schema'`,
so every top-level exported type from `schema.d.ts` is automatically
re-exported to consumers. Types inside the `_internal` namespace are still
reachable via `_internal.X` (the namespace itself is exported by the
wildcard) but their bare names are not.

The wildcard is load-bearing: it removes the maintenance burden of keeping
`lib/index.d.ts` in sync with new schema additions. If anyone ever swaps
it for an explicit allowlist, restore the per-name re-export verifier
that previously lived in `tasks/schema.mjs` (see git history) — otherwise

> new generated types will silently fail to surface in the public API.

Reviewer: Is this paragraph needed? I don't think it's likely someone would decide to remove the wildcard.

```suggestion
```

## CI integration

`npm run schema-typegen-diff-check` runs the generator and then verifies that
both `test/plot-schema.json` and `src/types/generated/` are unchanged via
`git diff --exit-code`. If either differs, exit code 1 — meaning either a
developer changed the source schema but didn't commit the regenerated
artifacts, or an attribute-file conversion silently altered the runtime

> schema and the change wasn't intentionally committed.

Reviewer: ```suggestion
`git diff --exit-code`. If either differs, the command fails with exit code 1
and outputs the diff to the console.
```
